src/v/wasm/allocator.cc

```diff
@@ -46,4 +52,76 @@ void heap_allocator::deallocate(heap_memory m) {
     _memory_pool.push_back(std::move(m));
 }
 
+stack_memory::stack_memory(stack_bounds bounds, allocated_memory data)
+  : _bounds(bounds)
+  , _data(std::move(data)) {}
+
+stack_memory::~stack_memory() {
+    if (!_data) {
+        // This can happen if the memory was moved.
+        return;
+    }
+    int r = ::mprotect(
+      _data.get(), _bounds.bottom - _data.get(), PROT_READ | PROT_WRITE);
+    vassert(r == 0, "stack memory must be able to unprotect on destruction");
+}
+
+size_t stack_memory::size() const { return _bounds.top - _bounds.bottom; }
+
+stack_bounds stack_memory::bounds() const { return _bounds; }
+
+stack_allocator::stack_allocator(config c)
+  : _tracking_enabled(c.tracking_enabled)
+  , _page_size(::getpagesize()) {}
+
+stack_memory stack_allocator::allocate(size_t size) {
+    size = ss::align_up(size, _page_size);
+    stack_memory mem;
+    // Reuse a page
+    if (!_memory_pool.empty() && _memory_pool.front().size() == size) {
+        mem = std::move(_memory_pool.front());
+        _memory_pool.pop_front();
+    } else {
+        // Create a stack with a guard page and aligned to a page.
+        auto buffer = ss::allocate_aligned_buffer<uint8_t>(
+          size + _page_size, _page_size);
+        // Protect the guard page by making it read only.
+        ::mprotect(buffer.get(), _page_size, PROT_READ);
```

Comment:
what's the motivation for even allowing reads?

Reply:
I'm just copying seastar's thread here: https://github.com/redpanda-data/seastar/blob/245e0ccfa6d58d7e0dca2b4034ce1bc43e39bdc5/src/core/thread.cc#L208

Interestingly they **don't** do that in release mode: https://github.com/redpanda-data/seastar/commit/1bdf0d8e041ce6b74a0880836e883364191fc877

Reply:
Also see [this conversation](https://redpandadata.slack.com/archives/C04B8659KGR/p1697641212503899?thread_ts=1697133282.506919&cid=C04B8659KGR) I had with Travis in slack.

Reply:
awesome discussion on slack, thanks